File: android/app/src/main/java/com/phicdy/totoanticipation/view/activity/GameListActivity.java
package com.phicdy.totoanticipation.view.activity;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.support.annotation.NonNull;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CompoundButton;
import android.widget.RadioButton;
import android.widget.TextView;

import com.phicdy.totoanticipation.R;
import com.phicdy.totoanticipation.model.Game;
import com.phicdy.totoanticipation.model.JLeagueRequestExecutor;
import com.phicdy.totoanticipation.model.JLeagueService;
import com.phicdy.totoanticipation.model.RakutenTotoRequestExecutor;
import com.phicdy.totoanticipation.model.RakutenTotoService;
import com.phicdy.totoanticipation.model.scheduler.DeadlineAlarm;
import com.phicdy.totoanticipation.model.storage.GameListStorage;
import com.phicdy.totoanticipation.model.storage.GameListStorageImpl;
import com.phicdy.totoanticipation.model.storage.SettingStorage;
import com.phicdy.totoanticipation.model.storage.SettingStorageImpl;
import com.phicdy.totoanticipation.presenter.GameListPresenter;
import com.phicdy.totoanticipation.view.GameListView;
import com.phicdy.totoanticipation.view.fragment.TeamInfoFragment;

import java.util.concurrent.Executor;

import fr.castorflex.android.smoothprogressbar.SmoothProgressBar;

public class GameListActivity extends AppCompatActivity implements GameListView {

    private boolean mTwoPane;
    private GameListPresenter presenter;
    private RecyclerView recyclerView;
    private SimpleItemRecyclerViewAdapter adapter;
    private SmoothProgressBar progressBar;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_game_list);

        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        toolbar.setTitle(getTitle());

        FloatingActionButton fab = (FloatingActionButton) findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                presenter.onFabClicked();
            }
        });

        recyclerView = (RecyclerView) findViewById(R.id.game_list);

        if (findViewById(R.id.item_detail_container) != null) {
            // The detail container view will be present only in the
            // large-screen layouts (res/values-w900dp).
            // If this view is present, then the
            // activity should be in two-pane mode.
            mTwoPane = true;
        }
        progressBar = (SmoothProgressBar) findViewById(R.id.progress);

        final RakutenTotoService rakutenTotoService = RakutenTotoService.Factory.create();
        final RakutenTotoRequestExecutor rakutenTotoRequestExecutor =
                new RakutenTotoRequestExecutor(rakutenTotoService);
        final JLeagueService jLeagueService = JLeagueService.Factory.create();
        final JLeagueRequestExecutor jLeagueRequestExecutor = new JLeagueRequestExecutor(jLeagueService);
        GameListStorage storage = new GameListStorageImpl(this);
        SettingStorage settingStorage = new SettingStorageImpl(this);
        presenter = new GameListPresenter(this, rakutenTotoRequestExecutor, jLeagueRequestExecutor, storage,
                settingStorage.isDeadlineNotify(), new DeadlineAlarm(this));
        presenter.onCreate();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.game_list, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_auto_anticipation:
                presenter.onOptionsAutoAnticipationSelected();
                break;
            case R.id.menu_setting:
                presenter.onOptionsSettingSelected();
                break;
        }
        return false;
    }

    @Override
    public void initList() {
        recyclerView.setVisibility(View.VISIBLE);
        adapter = new SimpleItemRecyclerViewAdapter();
        recyclerView.setAdapter(adapter);
    }

    @Override
    public void setTitleFrom(@NonNull String xxTh, @NonNull String deadline) {
        setTitle(getString(R.string.top_title, xxTh, deadline));
    }

    @Override
    public void startProgress() {
        progressBar.setVisibility(View.VISIBLE);
    }

    @Override
    public void stopProgress() {
        progressBar.progressiveStop();
        Handler handler = new Handler();
        handler.postDelayed(new Runnable() {
            @Override
            public void run() {
                progressBar.setVisibility(View.GONE);
            }
        }, 3000);
    }

    @Override
    public void startTotoAnticipationActivity(@NonNull String totoNum) {
        Intent intent = new Intent(this, TotoAnticipationActivity.class);
        intent.putExtra(TotoAnticipationActivity.KEY_TOTO_NUM, totoNum);
        startActivity(intent);
    }

    @Override
    public void goToSetting() {
        startActivity(new Intent(this, SettingActivity.class));
    }

    @Override
    public void showAnticipationStart() {
        View view = findViewById(android.R.id.content);
        if (view == null) return;
        Snackbar.make(view, R.string.start_auto_anticipation, Snackbar.LENGTH_SHORT).show();
    }

    @Override
    public void showAnticipationFinish() {
        View view = findViewById(android.R.id.content);
        if (view == null) return;
        Snackbar.make(view, R.string.finish_auto_anticipation, Snackbar.LENGTH_SHORT).show();
    }

    @Override
    public void notifyDataSetChanged() {
        adapter.notifyDataSetChanged();
    }

    class SimpleItemRecyclerViewAdapter
            extends RecyclerView.Adapter<SimpleItemRecyclerViewAdapter.ViewHolder> {

        @Override
        public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            View view = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.game_list_content, parent, false);
            return new ViewHolder(view);
        }

        @Override
        public void onBindViewHolder(final ViewHolder holder, final int position) {
            holder.game = presenter.gameAt(position);
            if (holder.game != null) {
                holder.tvHome.setText(getString(R.string.team_label, String.valueOf(holder.game.getHomeRanking()), holder.game.getHomeTeam()));
                holder.tvAway.setText(getString(R.string.team_label, String.valueOf(holder.game.getAwayRanking()), holder.game.getAwayTeam()));
            }

            holder.mView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (mTwoPane) {
                        Bundle arguments = new Bundle();
                        arguments.putString(TeamInfoActivity.ARG_HOME_TEAM, holder.game.getHomeTeam());
                        arguments.putString(TeamInfoActivity.ARG_AWAY_TEAM, holder.game.getAwayTeam());
                        TeamInfoFragment fragment = new TeamInfoFragment();
                        fragment.setArguments(arguments);
                        getSupportFragmentManager().beginTransaction()
                                .replace(R.id.item_detail_container, fragment)
                                .commit();
                    } else {
                        Context context = v.getContext();
                        Intent intent = new Intent(context, TeamInfoActivity.class);
                        intent.putExtra(TeamInfoActivity.ARG_HOME_TEAM, holder.game.getHomeTeam());
                        intent.putExtra(TeamInfoActivity.ARG_AWAY_TEAM, holder.game.getAwayTeam());

                        context.startActivity(intent);
                    }
                }
            });

            switch (holder.game.getAnticipation()) {
                case HOME:
                    holder.rbHome.setChecked(true);
                    break;
                case AWAY:
                    holder.rbAway.setChecked(true);
                    break;
                case DRAW:
                    holder.rbDraw.setChecked(true);
                    break;
            }
            holder.rbHome.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
                @Override
                public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                    presenter.onHomeRadioButtonClicked(holder.getAdapterPosition(), isChecked);
                }
            });
            holder.rbAway.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
                @Override
                public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                    presenter.onAwayRadioButtonClicked(holder.getAdapterPosition(), isChecked);
                }
            });
            holder.rbDraw.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
                @Override
                public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                    presenter.onDrawRadioButtonClicked(holder.getAdapterPosition(), isChecked);
                }
            });
        }

        @Override
        public int getItemCount() {
            return presenter.gameSize();
        }

        class ViewHolder extends RecyclerView.ViewHolder {
            final View mView;
            final TextView tvHome;
            final TextView tvAway;
            final RadioButton rbHome;
            final RadioButton rbAway;
            final RadioButton rbDraw;
            Game game;

            ViewHolder(View view) {
                super(view);
                mView = view;
                tvHome = (TextView) view.findViewById(R.id.tv_home);
                tvAway = (TextView) view.findViewById(R.id.tv_away);
                rbHome = (RadioButton) view.findViewById(R.id.rb_home);
                rbAway = (RadioButton) view.findViewById(R.id.rb_away);
                rbDraw = (RadioButton) view.findViewById(R.id.rb_draw);
            }

            @Override
            public String toString() {
                return super.toString() + " '" + tvAway.getText() + "'";
            }
        }
    }
}

File: android/app/src/main/java/com/phicdy/totoanticipation/view/activity/GameListActivity.kt
package com.phicdy.totoanticipation.view.activity

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.Handler
import android.text.Html
import android.text.method.LinkMovementMethod
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.RadioButton
import android.widget.TextView
import androidx.annotation.IntDef
import androidx.annotation.StringRes
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.android.material.snackbar.Snackbar
import com.phicdy.totoanticipation.BuildConfig
import com.phicdy.totoanticipation.R
import com.phicdy.totoanticipation.advertisement.AdProvider
import com.phicdy.totoanticipation.advertisement.AdViewHolder
import com.phicdy.totoanticipation.model.Game
import com.phicdy.totoanticipation.model.JLeagueRequestExecutor
import com.phicdy.totoanticipation.model.JLeagueService
import com.phicdy.totoanticipation.model.RakutenTotoRequestExecutor
import com.phicdy.totoanticipation.model.RakutenTotoService
import com.phicdy.totoanticipation.model.scheduler.DeadlineAlarm
import com.phicdy.totoanticipation.model.storage.GameListStorageImpl
import com.phicdy.totoanticipation.model.storage.SettingStorage
import com.phicdy.totoanticipation.model.storage.SettingStorageImpl
import com.phicdy.totoanticipation.presenter.GameListPresenter
import com.phicdy.totoanticipation.view.GameListView
import com.phicdy.totoanticipation.view.fragment.TeamInfoFragment
import dagger.android.support.DaggerAppCompatActivity
import fr.castorflex.android.smoothprogressbar.SmoothProgressBar
import javax.inject.Inject

class GameListActivity : DaggerAppCompatActivity(), GameListView {
    private var mTwoPane: Boolean = false
    private lateinit var presenter: GameListPresenter
    private lateinit var recyclerView: RecyclerView
    private lateinit var adapter: SimpleItemRecyclerViewAdapter
    private lateinit var progressBar: SmoothProgressBar

    @Inject
    lateinit var adProvider: AdProvider

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_game_list)

        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        toolbar.title = title

        val fab = findViewById<FloatingActionButton>(R.id.fab)
        fab.setOnClickListener { presenter.onFabClicked() }

        recyclerView = findViewById(R.id.game_list)

        if (findViewById<View>(R.id.item_detail_container) != null) {
            // The detail container view will be present only in the
            // large-screen layouts (res/values-w900dp).
            // If this view is present, then the
            // activity should be in two-pane mode.
            mTwoPane = true
        }
        progressBar = findViewById(R.id.progress)

        val rakutenTotoService = RakutenTotoService.Factory.create()
        val rakutenTotoRequestExecutor = RakutenTotoRequestExecutor(rakutenTotoService)
        val jLeagueService = JLeagueService.Factory.create()
        val jLeagueRequestExecutor = JLeagueRequestExecutor(jLeagueService)
        val storage = GameListStorageImpl(this)
        val settingStorage = SettingStorageImpl(this) as SettingStorage
        presenter = GameListPresenter(this, rakutenTotoRequestExecutor, jLeagueRequestExecutor, storage,
                DeadlineAlarm(this), settingStorage)
        presenter.onCreate()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.game_list, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_auto_anticipation -> presenter.onOptionsAutoAnticipationSelected()
            R.id.menu_setting -> presenter.onOptionsSettingSelected()
        }
        return false
    }

    override fun initList() {
        recyclerView.visibility = View.VISIBLE
        adapter = SimpleItemRecyclerViewAdapter()
        recyclerView.adapter = adapter
    }

    override fun setTitleFrom(xxTh: String, deadline: String) {
        title = getString(R.string.top_title, xxTh, deadline)
    }

    override fun startProgress() {
        progressBar.visibility = View.VISIBLE
    }

    override fun stopProgress() {
        progressBar.progressiveStop()
        val handler = Handler()
        handler.postDelayed({ progressBar.visibility = View.GONE }, 3000)
    }

    override fun startTotoAnticipationActivity(totoNum: String) {
        val intent: Intent
        if (BuildConfig.FLAVOR == "googlePlay") {
            // Google Play forbids to upload gambling app for Japan, open external browser
            val totoTopUrl = "http://sp.toto-dream.com/dci/sp/I/IMA/IMA01.do?op=inittotoSP&holdCntId=$totoNum"
            intent = Intent(Intent.ACTION_VIEW, Uri.parse(totoTopUrl))
        } else {
            intent = Intent(this, TotoAnticipationActivity::class.java)
            intent.putExtra(TotoAnticipationActivity.KEY_TOTO_NUM, totoNum)
        }
        startActivity(intent)
    }

    override fun goToSetting() {
        startActivity(Intent(this, SettingActivity::class.java))
    }

    override fun showAnticipationStart() {
        showSnackbar(R.string.start_auto_anticipation, Snackbar.LENGTH_SHORT)
    }

    override fun showAnticipationFinish() {
        showSnackbar(R.string.finish_auto_anticipation, Snackbar.LENGTH_SHORT)
    }

    override fun notifyDataSetChanged() {
        adapter.notifyDataSetChanged()
    }

    override fun showAnticipationNotSupport() {
        showSnackbar(R.string.anticipation_not_support, Snackbar.LENGTH_SHORT)
    }

    private fun showSnackbar(@StringRes res: Int, @SnackbarLength length: Int) {
        val view = findViewById<View>(android.R.id.content) ?: return
        Snackbar.make(view, res, length).show()
    }

    override fun showPrivacyPolicyDialog() {
        val alert = AlertDialog.Builder(this)
                .setMessage(Html.fromHtml(getString(R.string.privacy_policy_message)))
                .setCancelable(false)
                .setPositiveButton(R.string.accept) { _, _ ->
                    presenter.onPrivacyPolicyAccepted()
                }
                .setNegativeButton(R.string.not_accept) { _, _ ->
                    finish()
                }
                .create()
        alert.show()
        alert.findViewById<TextView>(android.R.id.message)?.movementMethod = LinkMovementMethod.getInstance()
    }

    @IntDef(Snackbar.LENGTH_SHORT, Snackbar.LENGTH_LONG)
    internal annotation class SnackbarLength

    internal inner class SimpleItemRecyclerViewAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

        private val VIEW_TYPE_CONTENT = 1
        private val VIEW_TYPE_AD = 2

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
            return when (viewType) {
                VIEW_TYPE_CONTENT -> {
                    val view = LayoutInflater.from(parent.context)
                            .inflate(R.layout.game_list_content, parent, false)
                    ViewHolder(view)
                }
                VIEW_TYPE_AD -> adProvider.newViewHolderInstance(parent)
                else -> throw IllegalStateException()
            }
        }

        override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
            when (holder) {
                is ViewHolder -> {
                    val game = presenter.gameAt(position)
                    holder.tvHome.text = when (game.homeRanking) {
                        Game.defaultRank -> getString(R.string.team_label, "- ", game.homeTeam)
                        else -> getString(R.string.team_label, game.homeRanking.toString(), game.homeTeam)
                    }
                    holder.tvAway.text = when (game.awayRanking) {
                        Game.defaultRank -> getString(R.string.team_label, "- ", game.awayTeam)
                        else -> getString(R.string.team_label, game.awayRanking.toString(), game.awayTeam)
                    }

                    holder.mView.setOnClickListener(View.OnClickListener { v ->
                        if (game.homeRanking == Game.defaultRank || game.awayRanking == Game.defaultRank) {
                            showSnackbar(R.string.not_support_foreign_league, Snackbar.LENGTH_SHORT)
                            return@OnClickListener
                        }
                        if (mTwoPane) {
                            val arguments = Bundle().apply {
                                putString(TeamInfoActivity.ARG_HOME_TEAM, game.homeTeam)
                                putString(TeamInfoActivity.ARG_AWAY_TEAM, game.awayTeam)
                            }
                            val fragment = TeamInfoFragment()
                            fragment.arguments = arguments
                            supportFragmentManager.beginTransaction()
                                    .replace(R.id.item_detail_container, fragment)
                                    .commit()
                        } else {
                            val context = v.context
                            val intent = Intent(context, TeamInfoActivity::class.java).apply {
                                putExtra(TeamInfoActivity.ARG_HOME_TEAM, game.homeTeam)
                                putExtra(TeamInfoActivity.ARG_AWAY_TEAM, game.awayTeam)
                            }

                            context.startActivity(intent)
                        }
                    })

                    when (game.anticipation) {
                        Game.Anticipation.HOME -> holder.rbHome.isChecked = true
                        Game.Anticipation.AWAY -> holder.rbAway.isChecked = true
                        Game.Anticipation.DRAW -> holder.rbDraw.isChecked = true
                    }
                    holder.rbHome.setOnCheckedChangeListener { _, isChecked -> presenter.onHomeRadioButtonClicked(holder.adapterPosition, isChecked) }
                    holder.rbAway.setOnCheckedChangeListener { _, isChecked -> presenter.onAwayRadioButtonClicked(holder.adapterPosition, isChecked) }
                    holder.rbDraw.setOnCheckedChangeListener { _, isChecked -> presenter.onDrawRadioButtonClicked(holder.adapterPosition, isChecked) }
                }
                is AdViewHolder -> holder.bind()
            }
        }

        override fun getItemCount(): Int {
            return presenter.gameSize() + 1 // 1 for Ad
        }

        override fun getItemViewType(position: Int) = if (position == presenter.gameSize()) VIEW_TYPE_AD else VIEW_TYPE_CONTENT
    }

    internal inner class ViewHolder(val mView: View) : RecyclerView.ViewHolder(mView) {
        val tvHome: TextView = mView.findViewById(R.id.tv_home)
        val tvAway: TextView = mView.findViewById(R.id.tv_away)
        val rbHome: RadioButton = mView.findViewById(R.id.rb_home)
        val rbAway: RadioButton = mView.findViewById(R.id.rb_away)
        val rbDraw: RadioButton = mView.findViewById(R.id.rb_draw)

        override fun toString(): String {
            return super.toString() + " '" + tvAway.text + "'"
        }
    }
}
